syntax = "proto3";

package person;

option java_multiple_files = true;
option java_package = "org.grpctest.java.common.define";
option java_outer_classname = "PersonProto";

// Pojos
message Person {
  optional int32 id = 1;
  optional string name = 2;
  optional int32 age = 3;
  optional string occupation = 4;
  optional Gender gender = 5;
  repeated string emails = 6;

  enum Gender {
    MALE = 0;
    FEMALE = 1;
    OTHER = 3;
  }
}

// Service messages and definition

message PersonId {
  optional int32 id = 1;
}

message GetPersonRequest {
  optional PersonId person_id = 1;
}

message GetPersonResponse {
  optional Person person = 1;
}

service PeopleService {
  // UNARY
  rpc GetPerson(GetPersonRequest) returns (GetPersonResponse);
  // SERVER_STREAMING
  rpc ListPerson(GetPersonRequest) returns (stream GetPersonResponse);
  // CLIENT_STREAMING
  rpc RegisterPerson(stream GetPersonRequest) returns (GetPersonResponse);
  // BIDI_STREAMING
  rpc StreamPerson(stream GetPersonRequest) returns (stream GetPersonResponse);
}
